### PR TITLE
Introduce SingleColumnTableType as DataType of SelectSymbol.

### PR DESCRIPTION
**Pull request by matriv**
This is is a preparation step to be able to support subselects
as value expressions in IN() and ANY().

### CODE REVIEW

File: sql/src/test/java/io/crate/integrationtests/SubSelectIntegrationTest.java [hunk 1]
```diff
@@ -314,6 +315,7 @@ public void testSingleRowSubselectWithMultipleRowsReturning() throws Exception {
     }
 
     @Test
+    @UseJdbc(0)
```
Comment (mikethebeer) [inline sql/src/test/java/io/crate/integrationtests/SubSelectIntegrationTest.java]: this is something that should not be part of this pr

File: sql/src/main/java/io/crate/operation/scalar/SingleValueFunction.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        if (arg0Value == null) {
+            return null;
+        }
+        switch (arg0Value.length) {
+            case 0:
+                return null;
+            case 1:
+                return arg0Value[0];
+            default:
+                throw new UnsupportedFeatureException("Subquery returned more than 1 row when it shouldn't.");
```
Comment (mikethebeer) [inline sql/src/main/java/io/crate/operation/scalar/SingleValueFunction.java]: Don't need to be a subquery at all. Would write `Query returned ...`

File: core/src/main/java/io/crate/types/SingleColumnTableType.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    public String getName() {
+        return innerType.getName() + "_table";
+    }
+
+    @Override
+    public Streamer<?> streamer() {
+        return this;
+    }
+
+    @Override
```
Comment (mikethebeer) [inline core/src/main/java/io/crate/types/SingleColumnTableType.java]: please document what it returns when calling `value` on the SCTT

File: core/src/main/java/io/crate/types/CollectionType.java [hunk 4]
```diff
@@ -21,14 +21,25 @@
...
+    public static DataType unnest(DataType dataType) {
         while (DataTypes.isCollectionType(dataType)) {
             dataType = ((CollectionType) dataType).innerType();
         }
         return dataType;
     }
+
+    @Override
+    public boolean isConvertableTo(DataType other) {
+        return other.id() == UndefinedType.ID ||
```
Comment (mikethebeer) [inline core/src/main/java/io/crate/types/CollectionType.java]: what if `other` is `null`?

File: sql/src/main/java/io/crate/operation/scalar/SingleValueFunction.java (hunk 2 above)
Comment (matriv) [inline sql/src/main/java/io/crate/operation/scalar/SingleValueFunction.java]: It's always a subquery, otherwise we don't have this restriction.

File: core/src/main/java/io/crate/types/CollectionType.java (hunk 4 above)
Comment (matriv) [inline core/src/main/java/io/crate/types/CollectionType.java]: When a value is null then the DataType = UndefinedType

File: core/src/main/java/io/crate/types/SingleColumnTableType.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    public Object[] value(Object value) {
+        if (value == null) {
+            return null;
+        }
+        return (Object[]) value;
+    }
+
+    @Override
+    public boolean isConvertableTo(DataType other) {
+        return super.isConvertableTo(other);
```
Comment (dobe) [inline core/src/main/java/io/crate/types/SingleColumnTableType.java]: why this override?

File: core/src/main/java/io/crate/types/SingleColumnTableType.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    @Override
+    public int compareValueTo(Object[] val1, Object[] val2) {
+        if (val1 == val2 || (val1 != null && Arrays.equals(val1, val2))) {
```
Comment (dobe) [inline core/src/main/java/io/crate/types/SingleColumnTableType.java]: do we need this override?

File: sql/src/main/java/io/crate/executor/transport/SingleRowSingleValueConsumer.java [hunk 7]
```diff
@@ -24,20 +24,22 @@
...
+import org.elasticsearch.common.util.iterable.Iterables;
 
+import java.util.ArrayList;
+import java.util.Collection;
 import java.util.stream.Collector;
 
 /**
  * BatchConsumer expects to receive only one row and triggers a future with the value once completed
  */
 class SingleRowSingleValueConsumer {
```
Comment (dobe) [inline sql/src/main/java/io/crate/executor/transport/SingleRowSingleValueConsumer.java]: is this still required?

File: sql/src/main/java/io/crate/operation/scalar/SingleValueFunction.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Override
+    public FunctionInfo info() {
+        return info;
+    }
+
+    static class SingleValueFunctionResolver extends BaseFunctionResolver {
+
+        SingleValueFunctionResolver() {
+            super(Signature.of(Signature.ArgMatcher.ANY_COLLECTION));
```
Comment (dobe) [inline sql/src/main/java/io/crate/operation/scalar/SingleValueFunction.java]: this should be only singlecoltabletype, right?

File: sql/src/main/java/io/crate/operation/scalar/SingleValueFunction.java (hunk 2 above)
Comment (mikethebeer) [inline sql/src/main/java/io/crate/operation/scalar/SingleValueFunction.java]: @matriv ok, understand

File: sql/src/main/java/io/crate/operation/operator/any/AnyEqOperator.java [hunk 9]
```diff
@@ -30,14 +30,17 @@
...
 
 public class AnyEqOperator extends AnyOperator {
 
     public static final String NAME = OPERATOR_PREFIX + ComparisonExpression.Type.EQUAL.getValue();
 
     public static FunctionInfo createInfo(DataType targetType) {
-        return new FunctionInfo(new FunctionIdent(NAME, ImmutableList.of(targetType, new SetType(targetType))), DataTypes.BOOLEAN);
+        return new FunctionInfo(
+            new FunctionIdent(NAME,
+                ImmutableList.of(targetType, new SingleColumnTableType(targetType))),
```
Comment (dobe) [inline sql/src/main/java/io/crate/operation/operator/any/AnyEqOperator.java]: discussed this already with @matriv 

this looks like this function only exists for singlecolumntabletype? is that right, what about 1 = any([1,2])? 

either we have here a public static method that is called only for one case of others, which is misleading - or we have no support for array types in the any operators and cast it?

File: sql/src/main/java/io/crate/operation/scalar/SingleValueFunction.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+                throw new UnsupportedFeatureException("Subquery returned more than 1 row when it shouldn't.");
+        }
+    }
+
+    @Override
+    public FunctionInfo info() {
+        return info;
+    }
+
+    static class MaxOneRowResolver extends BaseFunctionResolver {
```
Comment (dobe) [inline sql/src/main/java/io/crate/operation/scalar/SingleValueFunction.java]: pls rename

File: sql/src/main/java/io/crate/operation/scalar/SingleValueFunction.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+            super(Signature.of(Signature.ArgMatcher.ANY_SINGLE_COLUMN));
+        }
+
+        @Override
+        public FunctionImplementation getForTypes(List<DataType> dataTypes) throws IllegalArgumentException {
+            if (dataTypes.size() != 1) {
+                throw new IllegalArgumentException(NAME + " needs exactly one parameter");
+            }
+            DataType dataType = dataTypes.get(0);
+            if (!DataTypes.isCollectionType(dataType)) {
```
Comment (dobe) [inline sql/src/main/java/io/crate/operation/scalar/SingleValueFunction.java]: why accept any collection type here?

File: sql/src/main/java/io/crate/operation/scalar/SingleValueFunction.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+                throw new IllegalArgumentException("First parameter must be a collection type");
+            }
+            DataType returnType = ((SingleColumnTableType) dataType).innerType();
```
Comment (dobe) [inline sql/src/main/java/io/crate/operation/scalar/SingleValueFunction.java]: this is an unsafe cast, wondering why you dont get a warning

File: core/src/main/java/io/crate/types/CollectionType.java [hunk 13]
```diff
@@ -21,11 +21,98 @@
...
+/**
+ * A type which contains a collection of elements of another type.
+ */
+public abstract class CollectionType extends DataType implements Streamer {
+
+    protected DataType<?> innerType;
+
+    /**
+     * Construct a new Collection type
+     * @param innerType May be null if the type is read later through the Streamer interface
```
Comment (mxm) [inline core/src/main/java/io/crate/types/CollectionType.java]: Ok, will keep it for now. IMHO the empty constructor is not good practice since it prevents you from using `final` fields and is very error-prone if you forget to initialize. Will put this on my TODO list.

File: core/src/main/java/io/crate/types/DataTypes.java [hunk 14]
```diff
@@ -144,10 +150,11 @@
         .put(OBJECT.id(), ImmutableSet.of(GEO_SHAPE))
         .put(ArrayType.ID, ImmutableSet.of()) // convertability handled in ArrayType
         .put(SetType.ID, ImmutableSet.of()) // convertability handled in SetType
+        .put(SingleColumnTableType.ID, ImmutableSet.of()) // convertability handled in SetType
```
Comment (mxm) [inline core/src/main/java/io/crate/types/DataTypes.java]: I guess it is possible that a single column table might be cast to an array. This line doesn't add any support for casting since it returns an empty list.

File: core/src/main/java/io/crate/types/SingleColumnTableType.java [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+package io.crate.types;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+
+/**
+ * A type which represents a single column table. The type's
+ * value represents a collection of all single column rows.
```
Comment (mxm) [inline core/src/main/java/io/crate/types/SingleColumnTableType.java]: 👍

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 16]
```diff
@@ -774,9 +844,15 @@ protected Symbol visitSubqueryExpression(SubqueryExpression node, ExpressionAnal
...
-             * Since we only support 1 column and only single-row subselects it is okay to use the inner type directly.
+             * However, we support a single column RowType through the SingleColumnTableType.
              */
-            return new SelectSymbol(relation, fields.get(0).valueType());
+            Field field = fields.get(0);
+            SingleColumnTableType singleColumnTableType = new SingleColumnTableType(field.valueType());
+            SelectSymbol selectSymbol = new SelectSymbol(relation, singleColumnTableType);
+            return context.allocateFunction(
+                getBuiltinFunctionInfo(SingleValueFunction.NAME, Collections.singletonList(singleColumnTableType)),
+                // needs to be a mutable list...
```
Comment (mxm) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: Crate manipulates symbols in-place. This doesn't work if we use `Collections.singletonList(..)` here. Added a note.

File: sql/src/main/java/io/crate/executor/transport/SingleColumnConsumer.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.Collections;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.BinaryOperator;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collector;
+
+/**
+ * Collector to retrieve all values of a single column in a query.
```
Comment (mxm) [inline sql/src/main/java/io/crate/executor/transport/SingleColumnConsumer.java]: 👍

File: sql/src/main/java/io/crate/metadata/Scalar.java [hunk 18]
```diff
@@ -79,6 +80,8 @@
     public Symbol normalizeSymbol(Function symbol, TransactionContext transactionContext) {
         try {
             return evaluateIfLiterals(this, symbol);
+        } catch (UnsupportedFeatureException e) {
```
Comment (mxm) [inline sql/src/main/java/io/crate/metadata/Scalar.java]: There is a note in the commit message. We swallow all exceptions (even Throwables) upon evaluation of literals. This was tricky to debug and I would like to fix that to use checked exceptions instead. For now, I added a special type of exception to indicate that the function actually returned erroneously.

File: core/src/main/java/io/crate/types/SingleColumnTableType.java [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+        }
+        return (Object[]) value;
+    }
+
+    @Override
+    public Object[] readValueFrom(StreamInput in) throws IOException {
+        int size = in.readVInt();
+        if (size == 0) {
+            return null;
+        }
```
Comment (mxm) [inline core/src/main/java/io/crate/types/SingleColumnTableType.java]: `serde`? Will create a unit test and pull up the serialization code.

File: sql/src/main/java/io/crate/operation/operator/any/AnyEqOperator.java (hunk 9 above)
Comment (mxm) [inline sql/src/main/java/io/crate/operation/operator/any/AnyEqOperator.java]: This was using `SetType` before which was also not correct. I think using the `ArrayType` should be ok. We can cast the `SingleColumnTableType` to `ArrayType`.

File: core/src/main/java/io/crate/types/DataTypes.java (hunk 14 above)
Comment (dobe) [inline core/src/main/java/io/crate/types/DataTypes.java]: hm, why is it then in the map at all?

File: core/src/main/java/io/crate/types/SingleColumnTableType.java (hunk 19 above)
Comment (dobe) [inline core/src/main/java/io/crate/types/SingleColumnTableType.java]: serde = serialization / deserialization

File: sql/src/main/java/io/crate/metadata/Scalar.java (hunk 18 above)
Comment (dobe) [inline sql/src/main/java/io/crate/metadata/Scalar.java]: pls add the note in the code, no one is looking at the commit message afterwards ... you mean you changed if for debugging purposes? wondering why this is actually not a change in contract. probably evaluation of literals must not throw exceptions at all?

File: sql/src/main/java/io/crate/operation/operator/any/AnyEqOperator.java (hunk 9 above)
Comment (dobe) [inline sql/src/main/java/io/crate/operation/operator/any/AnyEqOperator.java]: there are 2, array and singlecoltabletype

File: core/src/main/java/io/crate/types/CollectionType.java [hunk 20]
```diff
@@ -38,12 +38,18 @@
...
      * Construct a new Collection type
-     * @param innerType May be null if the type is read later through the Streamer interface
+     * @param innerType The type of the elements inside the collection
      */
     public CollectionType(DataType<?> innerType) {
         this.innerType = innerType;
     }
 
+    /**
+     * Constructor used for the Streamer interface which initializes
```
Comment (dobe) [inline core/src/main/java/io/crate/types/CollectionType.java]: iface is called Streamable maybe link?

File: sql/src/main/java/io/crate/executor/transport/FirstColumnConsumer.java [hunk 21]
```diff
@@ -36,9 +36,9 @@
 import java.util.stream.Collector;
 
 /**
- * Collector to retrieve all values of a single column in a query.
+ * Collector to retrieve all values of the first column in a query.
```
Comment (dobe) [inline sql/src/main/java/io/crate/executor/transport/FirstColumnConsumer.java]: sorry did not see it in the first place: but it takes the first column of each row. (not query)

File: sql/src/main/java/io/crate/metadata/Scalar.java (hunk 18 above)
Comment (mxm) [inline sql/src/main/java/io/crate/metadata/Scalar.java]: I already added the comment in addition to the commit message. You added your GitHub comment above it. Btw, I find it very helpful to look inside commit messages through the IntelliJ's annotate function to add a bigger picture to changes.

Comment (mxm) [inline sql/src/main/java/io/crate/metadata/Scalar.java]: > you mean you changed if for debugging purposes?

No, not for debugging purposes but the existing code can produce hard-to-debug errors because it swallows arbitrary exceptions. This is definitely not good practice.

File: sql/src/main/java/io/crate/operation/scalar/SingleValueFunction.java (hunk 12 above)
Comment (dobe) [inline sql/src/main/java/io/crate/operation/scalar/SingleValueFunction.java]: the cast is still unchecked?

File: sql/src/main/java/io/crate/executor/transport/FirstColumnConsumer.java (hunk 21 above)
Comment (mxm) [inline sql/src/main/java/io/crate/executor/transport/FirstColumnConsumer.java]: 👍

File: core/src/main/java/io/crate/types/CollectionType.java (hunk 20 above)
Comment (mxm) [inline core/src/main/java/io/crate/types/CollectionType.java]: Generally this is a problem of the `Streamable` interface but here it is Crate's `Streamer`.

Comment (mxm) [inline core/src/main/java/io/crate/types/CollectionType.java]: Added a link

**Comment by mxm** [comment]
@dobe @mibe I've addressed your comments in the following commits:

937e3e1 fixup! move note to inner method
c59d839 fixup! cosmetic cleanup in equals
995c19c fixup! revert back to Streamer empty constructor
a151b98 fixup! correct comment
7b34d33 fixup! correct doc string
a2643fc fixup! explain mutable list comment
77434f6 fixup! rename SingleColumnConsumer to FirstColumnConsumer
55d65e2 fixup! add note about swallowed exception
acd241d fixup! rename resolver
8c2b660 fixup! narrow type check in single value function
ce0cc01 fixup! remove createInfo on AnyEqOperator
bfaa0e9 fixup! override normalizeSymbol, throw UnsupportedOperationException instead of UnsupportedFeatureException
a8b525e fixup! adjust javadoc
2524658 fixup! Handle null values correctly / decouple streamer implementation
ba2d5ec fixup! add TableType serialization test
e4ba476 fixup! add test for SingleValueFunction

File: core/src/main/java/io/crate/types/CollectionType.java [hunk 22]
```diff
@@ -45,7 +45,7 @@ public CollectionType(DataType<?> innerType) {
     }
 
     /**
-     * Constructor used for the Streamer interface which initializes
+     * Constructor used for the {@link Streamer} interface which initializes
```
Comment (dobe) [inline core/src/main/java/io/crate/types/CollectionType.java]: it is org.elasticsearch.common.io.stream.Streamable there is also a  reasoning about the ctor

File: sql/src/test/java/io/crate/operation/scalar/SingleValueFunctionTest.java [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.crate.operation.scalar;
+
+public class SingleValueFunctionTest {
```
Comment (dobe) [inline sql/src/test/java/io/crate/operation/scalar/SingleValueFunctionTest.java]: remove?

File: core/src/main/java/io/crate/types/CollectionType.java (hunk 22 above)
Comment (mxm) [inline core/src/main/java/io/crate/types/CollectionType.java]: Oh yes, of course it is `Streamable` and not `Streamer`. We implement both here but the ctor affects `Streamable`.
